webapp/src/monacoEditAmendments.ts

```diff
@@ -0,0 +1,168 @@
+import { rangeToSelection } from "./monaco";
+
+// This file adds a feature to Monaco I (dz) call "Edit Amendments".
+// EditAmendments are used by the auto-complete system as a work around for limitations in Monaco.
+// Worked example:
+//      Consider if a user starts typing this (in Minecraft py):
+//          player.on_cha
+//      Our auto-complete suggestion would be:
+//          player.on_chat
+//      And the snippet we want to insert is:
+//          def on_chat_handler():
+//              pass
+//          player.on_chat(on_chat_handler)
+//      The problem is that Monaco would paste this snippet as:
+//          player.def on_chat_handler():
+//              pass
+//          player.on_chat(on_chat_handler)
+//      And there's no way I've found to remove the "player." prefix.
+//      Note that we don't have this issue in typescript because of inline callbacks:
+//          player.on_chat(() => {
+//          
+//          })
+//      So the system basically sticks a json payload at the end of the snippet insertion:
+//          player.on_chat#AMENDMENT{behavior: "replaceLine", insertText: "def on_chat_handler(): ..." }
+//      which we then sniff for in the onDidChangeModelContent event and apply the changes.
+// Known limitations:
+//  - Any time we use an edit amendment, the user's undo stack will have a superfluous entry.
+//    For example, the undo stack for player.on_chat will look like:
+//      After edit:
+//          def on_chat_handler():
+//              pass
+//          player.on_chat(on_chat_handler)
+//      1st undo:
+//          player.
+//      2nd undo:
+//          player.on_cha
+//    We could remove this extra entry if Monaco gave us a way to pop items from the undo stack,
+//    or better yet if it supported our scenarios so that we didn't need edit amendments.
+//  - The user may see a flicker of "#AMENDMENT:..." after completing a snippet
+// Possible future issues:
+//  - The implementation depends on using setTimeout() as a way to wait for Monaco, because
+//    onDidChangeModelContent event notifies us before the change is applied to the buffer
+//    presumably as a way to give us a chance to react somehow, however to apply another edit
+//    we have to wait for the first edit to be applied, thus we use setTimeout().
+//    However in most cases when you use setTimeout(), it's likely that you may be susceptible to
+//    race conditions, especially if anyone else is listening onDidChangeModelContent and using
+//    setTimeout().
+
+export interface EditAmendment {
+    behavior: "replaceLine",
+    insertText: string,
+    selectText?: string,
+}
+type EditAmendmentInstance = {
```

Comment:
what is this weird syntax?

Reply:
It's semantically (nearly) the same as:

interface EditAmendmentInstance extends EditAmendement {
   range: monaco.IRange
}

I can change it. It reads slightly differently, it's saying: "An EditAmendmentInstance has a range and everything that an EditAmendment has".

Reply:
I'll change it, interfaces are more idiomatic

Reply:
![image](https://user-images.githubusercontent.com/6453828/67591002-35c33100-f711-11e9-9e05-fdaa95a5a2d7.png)

Reply:
My brain likes to think in terms of algebraic types instead of inheritance :)